storage/innobase/lock/lock0lock.cc

```diff
@@ -2376,10 +2549,14 @@ static void lock_rec_dequeue_from_page(lock_t *in_lock, bool owns_wait_mutex)
 	the first X lock that is waiting or has been granted. */
 
 	for (lock_t* lock = lock_sys_t::get_first(cell, page_id);
-	     lock != NULL;
-	     lock = lock_rec_get_next_on_page(lock)) {
-
+	     lock != NULL;) {
+		/* Store pointer to the next element, because if some lock is
+		bypassed, the pointer to the next lock in the current lock
+		object will be changed, as the current lock will change
+		its position in lock queue. */
+		lock_t *next= lock_rec_get_next_on_page(lock);
 		if (!lock->is_waiting()) {
+			lock= next;
 			continue;
 		}
```

Comment:
Please apply my suggestion for `lock_rec_rebuild_waiting_queue()`.

In fact, these two functions look rather similar. Would it be feasible to unify them somehow, maybe using some `template` parameter to reduce source code duplication?